### PR TITLE
45 logging levels

### LINKED ISSUE #45: Specify level of debug information
After merging #25 the CLI now spits out a bunch of debug information.  This debug information should not be displayed when running the CLI unless specifically asked for.  For example running the CLI with no arguments produces:

```
> Corgibytes.Freshli.Cli.exe
info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Production
info: Microsoft.Hosting.Lifetime[0]
      Content root path: C:\Repos\freshli-cli\Corgibytes.Freshli.Cli\bin\Debug\net5.0
[Command Execution Invocation Started  - ParseResult: ![ Corgibytes.Freshli.Cli ] ]
[Command Execution Invocation Ended - ParseResult: ![ Corgibytes.Freshli.Cli ] ]
info: Microsoft.Hosting.Lifetime[0]
      Application is shutting down...
Required command was not provided.
Corgibytes.Freshli.Cli
  Root Command

Usage:
  Corgibytes.Freshli.Cli [options] [command]

Options:
  -?, -h, --help  Show help and usage information


Commands:
  scan <path>  Scan command returns metrics results for given local repository path
```

The `info:` and `[Command Execution Invocation]` should not be displayed when running the application.  You will see similar `info` lines when running the `scan` command:

```
> Corgibytes.Freshli.Cli.exe scan https://github.com/corgibytes/freshli-fixture-ruby-nokotest.git
info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Production
info: Microsoft.Hosting.Lifetime[0]
      Content root path: C:\Repos\freshli-cli\Corgibytes.Freshli.Cli\bin\Debug\net5.0
[Command Execution Invocation Started  - ParseResult: [ Corgibytes.Freshli.Cli [ scan <https://github.com/corgibytes/freshli-fixture-ruby-nokotest.git> *[ --format <Json> ] *[ --output <Console> ] ] ] ]
CliOutput.ScanCommand_ScanCommand_Executing_scan_command_handler
Sending metrics to Console
[
  {
    "Filename": "Gemfile.lock",
    "MetricResults": [
      {
        "Date": "2017-01-01T23:59:59.9999999",
        "LibYear": [
          {
 <... More JSON...>
```

The main issue to remove the debug information when running the application.  Time permitting it would be nice to add a `verbose` argument.  Something like:

```
 Corgibytes.Freshli.Cli.exe  -verbose scan https://github.com/corgibytes/freshli-fixture-ruby-nokotest.git
```

This argument would be apply to all commands and display the `info` information.  If need be we can also set a verbose level, something like:

```
 Corgibytes.Freshli.Cli.exe -verbose [debug|info] scan https://github.com/corgibytes/freshli-fixture-ruby-nokotest.git
```

The above are just examples, feel free to suggest better ways to display debug/verbose information.

### LINKED ISSUE #175: Exceptions thrown by `IApplicationActivity.Handle` and `IApplicationEvent.Handle` need to be more visible
We recently had a problem where an `IApplicationActivity.Handle` method was throwing an exception, but no message was getting written to the console. Since we don't have a log file in place yet (we still have some clean-up work to do with #45) the exception and stack trace need to be output to the console. Once #45 gets merged in, then we can have the full stack trace go to the log and the error message from the exception can be displayed on the console.

### PR DESCRIPTION
**Pull request by lizcorgi**
Fixes #45.
Fixes #175.

This correctly configures the application to use NLog with a default level of Warn for all logs in the application, including those from Microsoft.Hosting.

The log level can be modified with a commandline option. The log can be sent to a file instead with a commandline option.

Many classes where explicitly writing to console for logs. These were updated to use a logger so that all logs will follow these settings. The only thing that should be explicitly written to the console should be the output of the scan.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -3,33 +3,33 @@
...
+1. Download it.
+2. Extract the Zip file.
+3. Copy the folder to an apporiate location.
+4. Open a terminal and run `Corgibytes.Freshli.Cli.exe`.
 
 To run Freshli:
 
 ```
-Path\To\Freshli\Corgibytes.Freshli.Cli.exe <repo>
+Path\To\Freshli\Corgibytes.Freshli.Cli.exe scan <repo>
```
Comment (mrbiggred) [inline README.md:20]: Thank you for adding the `scan` command to the ReadMe.

File: README.md [hunk 2]
```diff
@@ -174,23 +173,37 @@ If you have v1.0.1 instealled then your libyear when checking on May 1, 2019 is
...
+Log levels can be adjusted by using the `--logLevel <level>` option when running the application. The level can be any level that is supported by NLog:
+
+-   Trace
+-   Debug
+-   Info
+-   Warn
+-   Error
+-   Fatal
+
+Logs can be redirected to a file instead by using the `--logFile <file_path_and_name>` option when running the application.
```
Comment (mrbiggred) [inline README.md:203]: Can you add an example of a command line that uses the `logLevel` and `logFile` options?

File: Corgibytes.Freshli.Cli/Program.cs [hunk 3]
```diff
@@ -24,25 +31,66 @@ public static async Task<int> Main(string[] args)
...
+                .ConfigureServices((_, services) =>
             {
                 new FreshliServiceBuilder(services).Register();
             });
 
         public static CommandLineBuilder CreateCommandLineBuilder()
         {
+            Option logLevelOption = new Option<string>(alias: "--loglevel", description:"the minimum log level to log to console",
+                getDefaultValue: () => "Warn");
+            Option logFileOption = new Option<string?>("--logfile", "file for logs instead of logging to console");
```
Comment (mrbiggred) [inline Corgibytes.Freshli.Cli/Program.cs:75]: I think the `logLevelOption` and `logFileOption` should be classes in the `CommandOptions/` folder.  Like the `ScanCommandOptions` class.  Can you check with @mfcorgi?

File: Corgibytes.Freshli.Cli.Test/ProgramTest.cs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+            stringWriter.ToString().Should().Contain("DEBUG|Microsoft.Extensions.Hosting.Internal.Host:0|Hosting stopping");
+        }
+
+        [Fact]
+        public void Validate_Main_loglevel_info()
+        {
+            StringWriter stringWriter = new StringWriter();
+            Console.SetOut(stringWriter);
+            Program.Main(new string [] {"--loglevel", "Info" });
+            stringWriter.ToString().Should().NotContain("DEBUG|Microsoft.Extensions.Hosting.Internal.Host:0|Hosting stopping");
```
Comment (mrbiggred) [inline Corgibytes.Freshli.Cli.Test/ProgramTest.cs:29]: Why testing `NotContain`?

File: Corgibytes.Freshli.Cli.Test/ProgramTest.cs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+{
+    public class ProgramTest : FreshliTest
+    {
+        public ProgramTest(ITestOutputHelper output) : base(output) {}
+
+        [Fact]
+        public void Validate_Main_loglevel_debug()
+        {
+            StringWriter stringWriter = new StringWriter();
+            Console.SetOut(stringWriter);
```
Comment (mrbiggred) [inline Corgibytes.Freshli.Cli.Test/ProgramTest.cs:18]: Could some of the repeated code at the start of each test method be moved to a setup function?

File: Corgibytes.Freshli.Cli/Program.cs [hunk 6]
```diff
@@ -24,25 +31,66 @@ public static async Task<int> Main(string[] args)
...
+                FileTarget fileTarget = new FileTarget();
+                config.AddTarget("file", fileTarget);
+                fileTarget.FileName = logfile;
+                fileTarget.Layout = DefaultLogLayout;
+                config.LoggingRules.Add(new LoggingRule("*", desiredLevel, fileTarget));
+            }
+
+            LogManager.Configuration = config;
+        }
+
```
Comment (mrbiggred) [inline Corgibytes.Freshli.Cli/Program.cs:57]: What do you think about moving the logger setup to it's own class?

**Review by mrbiggred — DISMISSED** [review]
My feedback is mostly checking if the command line options and the logic to setup the logging should be moved to their own class.  I know @mfcorgi has a folder that contains some existing command options and I wonder if the logging options should go there as well.

Also when I run the automated tests in my local environment, Rider on Windows 10, one test  fails.  It is the `ProgramTests.Validate_Main_Logfile` and fails with the following below error.  Any idea why?

```
System.IO.IOException: The process cannot access the file 'C:\Repos\freshli-cli\Corgibytes.Freshli.Cli.Test\bin\Debug\net5.0\testlog.log' because it i...

System.IO.IOException
The process cannot access the file 'C:\Repos\freshli-cli\Corgibytes.Freshli.Cli.Test\bin\Debug\net5.0\testlog.log' because it is being used by another process.
   at System.IO.FileStream.ValidateFileHandle(SafeFileHandle fileHandle)
   at System.IO.FileStream.CreateFileOpenHandle(FileMode mode, FileShare share, FileOptions options)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options)
   at System.IO.StreamReader.ValidateArgsAndOpenPath(String path, Encoding encoding, Int32 bufferSize)
   at System.IO.StreamReader..ctor(String path, Encoding encoding, Boolean detectEncodingFromByteOrderMarks)
   at System.IO.File.InternalReadAllText(String path, Encoding encoding)
   at System.IO.File.ReadAllText(String path)
   at Corgibytes.Freshli.Cli.Test.ProgramTest.Validate_Main_logfile() in C:\Repos\freshli-cli\Corgibytes.Freshli.Cli.Test\ProgramTest.cs:line 51
```

Thank you for updating the ReadMe with the `scan` command and also adding documentation on how to use the log levels.

**Comment by mscottford** [comment]
Over the next couple of weeks, I'm going to be working on reviving this pull request to get this functionality merged in.

File: Corgibytes.Freshli.Cli.Test/ProgramTest.cs (hunk 4 above)
Comment (mscottford) [reply Corgibytes.Freshli.Cli.Test/ProgramTest.cs]: This test is making sure that the logger does not produce any `DEBUG` messages when the log level is set to `INFO`.

File: README.md (hunk 2 above)
Comment (mscottford) [reply README.md:218]: Issue #214 has been created to track the need to update the README with this information.

File: Corgibytes.Freshli.Cli/Program.cs (hunk 3 above)
Comment (mscottford) [reply Corgibytes.Freshli.Cli/Program.cs]: The `ScanCommandOptions` class is used by the parameter binding functionality in `System.CommandLine` to pull the parameter values out of the parsed string and store them in objects. We don't have the opportunity to do that binding here (at least I don't think we do). Instead the values are being looked up dynamically. All of that to say that I don't think it makes sense to create a separate `CommandOptions` class for `MainCommand`.

Comment (mscottford) [reply Corgibytes.Freshli.Cli/Program.cs]: What I did do was move some of this setup into the `MainCommand` class. That helps consolidate that logic and make this chunk of code a little more readable.

File: Corgibytes.Freshli.Cli/Program.cs (hunk 6 above)
Comment (mscottford) [reply Corgibytes.Freshli.Cli/Program.cs:58]: I created #219 to track this request.

File: Corgibytes.Freshli.Cli.Test/ProgramTest.cs (hunk 5 above)
Comment (mscottford) [reply Corgibytes.Freshli.Cli.Test/ProgramTest.cs]: This has been addressed in [2709615](https://github.com/corgibytes/freshli-cli/pull/46/commits/270961586a28c31c80392046fc7d011334a35e8d).

File: Corgibytes.Freshli.Cli/Extensions/ParseResultExtensions.cs [hunk 7]
```diff
@@ -31,8 +31,17 @@ public static class ParseResultExtensions
...
+    public static T? GetOptionValueByName<T>(this ParseResult result, string name)
+    {
+        try
+        {
+            return result.FindOptionAndGetValue<T>(x => x.Name.Equals(name));
+        }
+        catch (InvalidOperationException error)
+        {
+            throw new ArgumentException($"No option was found with the name `{name}`.", error);
+        }
```
Comment (edwinkortman) [inline Corgibytes.Freshli.Cli/Extensions/ParseResultExtensions.cs:51]: Yay! No more "could not find option in sequence" confusion.

**Review by edwinkortman — APPROVED** [review]
Awesome improvements!

**Pull request merged**